### PR TITLE
Add yolov3 model

### PR DESCRIPTION
**Pull request by jiafatom**
# YOLOv3

## Description
This model is a neural network for real-time object detection that detects 80 different classes. It is very fast and accurate.

## Model

|Model        |Download  |Checksum|Download (with sample test data)|ONNX version|Opset version|Accuracy |
|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|:--------------|
|YOLOv3       |[237 MB](https://onnxzoo.blob.core.windows.net/models/opset_10/yolov3/yolov3.onnx) | [MD5](https://onnxzoo.blob.core.windows.net/models/opset_10/yolov3/yolov3-md5.txt) |[222 MB](https://onnxzoo.blob.core.windows.net/models/opset_10/yolov3/yolov3.tar.gz)|1.5 |10 |mAP of 0.553 |

<hr>

### Input to model
Resized image `(1x3x416x416)`
Original image size `(1x2)` which is `[image.size[1], image.size[0]]` 

### Preprocessing steps
The images have to be loaded in to a range of [0, 1]. The transformation should preferrably happen at preprocessing.

The following code shows how to preprocess a NCHW tensor:

```python
import numpy as np
from PIL import Image

# this function is from yolo3.utils.letterbox_image
def letterbox_image(image, size):
    '''resize image with unchanged aspect ratio using padding'''
    iw, ih = image.size
    w, h = size
    scale = min(w/iw, h/ih)
    nw = int(iw*scale)
    nh = int(ih*scale)

    image = image.resize((nw,nh), Image.BICUBIC)
    new_image = Image.new('RGB', size, (128,128,128))
    new_image.paste(image, ((w-nw)//2, (h-nh)//2))
    return new_image

def preprocess(img):
    model_image_size = (416, 416)
    boxed_image = letterbox_image(img, tuple(reversed(model_image_size)))
    image_data = np.array(boxed_image, dtype='float32')
    image_data /= 255.
    image_data = np.transpose(image_data, [2, 0, 1])
    image_data = np.expand_dims(image_data, 0)
    return image_data

image = Image.open(img_path)
# input
image_data = preprocess(image)
image_size = np.array([image.size[1], image.size[0]], dtype=np.int32).reshape(1, 2)
```

### Output of model
The model has 3 outputs.
boxes: `(1x'n_candidates'x4)`, the coordinates of all anchor boxes,
scores: `(1x80x'n_candidates')`, the scores of all anchor boxes per class,
indices: `('nbox'x3)`, selected indices from the boxes tensor. The selected index format is (batch_index, class_index, box_index). The class list is [here](https://github.com/qqwweee/keras-yolo3/blob/master/model_data/coco_classes.txt)

## Postprocessing steps
Post processing and meaning of output
```
out_boxes, out_scores, out_classes = [], [], []
for idx_ in indices:
    out_classes.append(idx_[1])
    out_scores.append(scores[tuple(idx_)])
    idx_1 = (idx_[0], idx_[2])
    out_boxes.append(boxes[idx_1])
```
out_boxes, out_scores, out_classes are list of resulting boxes, scores, and classes.
<hr>

## Dataset (Train and validation)
We use pretrained weights from pjreddie.com [here](https://pjreddie.com/media/files/yolov3.weights).
<hr>

## Validation accuracy
Metric is COCO box mAP (averaged over IoU of 0.5:0.95), computed over 2017 COCO val data.
mAP of 0.553 based on original Yolov3 model [here](https://pjreddie.com/darknet/yolo/)
<hr>

## Publication/Attribution
Joseph Redmon, Ali Farhadi. YOLOv3: An Incremental Improvement, [paper](https://pjreddie.com/media/files/papers/YOLOv3.pdf)

<hr>

## References
This model is converted from a keras model [repository](https://github.com/qqwweee/keras-yolo3) using keras2onnx convertor [repository](https://github.com/onnx/keras-onnx).
<hr>

## License
MIT License
<hr>

### CODE REVIEW

File: yolov3/README.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+indices: `('nbox'x3)`, selected indices from the boxes tensor. The selected index format is (batch_index, class_index, box_index).
+
+Here is the python API to run onnx model:
+```
+import onnxruntime
+sess = onnxruntime.InferenceSession('yolov3.onnx')
+feed_f = dict(zip(['input_1:01', 'image_shape:01'],
+             (image_data, np.array([image.size[1], image.size[0]], dtype=np.int32).reshape(1, 2))))
+all_boxes, all_scores, indices = sess.run(None, input_feed=feed_f)
+```
```
Comment (ebarsoum) [inline yolov3/README.md]: This is ONNX Runtime, the above code should move to ORT repo and put a link to it here.

File: yolov3/README.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+Joseph Redmon, Ali Farhadi. YOLOv3: An Incremental Improvement, [paper](https://pjreddie.com/media/files/papers/YOLOv3.pdf)
+
+<hr>
+
+## References
+This model is converted from a keras model [repository](https://github.com/qqwweee/keras-yolo3) using
+keras2onnx convertor [repository](https://github.com/onnx/keras-onnx).
+<hr>
+
+## License
```
Comment (prasanthpul) [inline yolov3/README.md]: Can you double check this is the right license?

File: yolov3/README.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
+# YOLOv3
+
+## Description
+This model is a real-time neural network for object detection that detects 80 different classes.
```
Comment (prasanthpul) [inline yolov3/README.md]: this looks like description for SSD. is it accurate for Yolov3?

File: yolov3/README.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+## Model
+
+|Model        |Download  |Checksum|ONNX version|Opset version|Accuracy |
+|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|
+|YOLOv3       |[237 MB](https://onnxzoo.blob.core.windows.net/models/opset_10/yolov3/yolov3.onnx) | [MD5](https://onnxzoo.blob.core.windows.net/models/opset_10/yolov3/yolov3-md5.txt) |1.4.1 |10 |mAP of 0.553 |
```
Comment (prasanthpul) [inline yolov3/README.md]: is it opset 10 or onnx 1.4.1?

File: yolov3/README.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    image_data /= 255.
+    image_data = np.transpose(image_data, [2, 0, 1])
+    image_data = np.expand_dims(image_data, 0)
+    return image_data
+```
+
+### Output of model
+The model has 3 outputs.
+boxes: `(1x'n_candidates'x4)`, the coordinates of all anchor boxes,
+scores: `(1x80x'n_candidates')`, the scores of all anchor boxes per class,
```
Comment (prasanthpul) [inline yolov3/README.md]: where is the list of classes?

File: yolov3/README.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+```
+out_boxes, out_scores, out_classes are list of resulting boxes, scores, and classes.
+<hr>
+
+## Dataset (Train and validation)
+We use pretrained weights from pjreddie.com [here](https://pjreddie.com/media/files/yolov3.weights).
+<hr>
+
+## Validation accuracy
+Metric is COCO box mAP (averaged over IoU of 0.5:0.95), computed over 2017 COCO val data.
```
Comment (prasanthpul) [inline yolov3/README.md]: These metrics are based on the ONNX model or the original Yolov3 model?

File: yolov3/README.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    out_classes.append(idx_[1])
+    out_scores.append(all_scores[tuple(idx_)])
+    idx_1 = (idx_[0], idx_[2])
+    out_boxes.append(all_boxes[idx_1])
+```
+out_boxes, out_scores, out_classes are list of resulting boxes, scores, and classes.
+<hr>
+
+## Dataset (Train and validation)
+We use pretrained weights from pjreddie.com [here](https://pjreddie.com/media/files/yolov3.weights).
```
Comment (prasanthpul) [inline yolov3/README.md]: Is there a notebook showing training we can link to?

File: yolov3/README.md (hunk 2 above)
Comment (jiafatom) [reply yolov3/README.md]: The license for yolov3 keras model we used, right? It is MIT license, 
https://github.com/qqwweee/keras-yolo3/blob/master/LICENSE
will fix.

File: yolov3/README.md (hunk 3 above)
Comment (jiafatom) [reply yolov3/README.md]: Yes. It is a real-time object detector.

File: yolov3/README.md (hunk 4 above)
Comment (jiafatom) [reply yolov3/README.md]: opset 10, onnx 1.5, will fix.

File: yolov3/README.md (hunk 5 above)
Comment (jiafatom) [reply yolov3/README.md]: The 3rd output is the indices from the boxes tensor, I.e., (batch_index, class_index, box_index). So it contains class_index. The list of classes is coco_classes. we can find it here:
https://github.com/qqwweee/keras-yolo3/blob/master/model_data/coco_classes.txt
will add a link here.

File: yolov3/README.md (hunk 6 above)
Comment (jiafatom) [reply yolov3/README.md]: Based on original Yolov3 model reported on their website.

File: yolov3/README.md (hunk 7 above)
Comment (jiafatom) [reply yolov3/README.md]: We used a standard pretrain weights provided by yolov3. The convertor team is working on the jupyter notebook about end-to-end pipeline. Training of original keras model is here: https://github.com/qqwweee/keras-yolo3
We need finish the notebook and then add training. Currently it is fine we just put the pretrained model to model zoo.

File: yolov3/README.md (hunk 1 above)
Comment (jiafatom) [reply yolov3/README.md]: This is standard method to call ORT, I will skip this. I will put example input/output data and upload the tar.gz files for the user to use. Will add it in the table called: Download (with sample test data), this way it is better for illustration rather than link to ORT.

File: yolov3/README.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+image_size = np.array([image.size[1], image.size[0]], dtype=np.int32).reshape(1, 2)
+```
+
+### Output of model
+The model has 3 outputs.
+boxes: `(1x'n_candidates'x4)`, the coordinates of all anchor boxes,
+scores: `(1x80x'n_candidates')`, the scores of all anchor boxes per class,
+indices: `('nbox'x3)`, selected indices from the boxes tensor. The selected index format is (batch_index, class_index, box_index). The class list is [here](https://github.com/qqwweee/keras-yolo3/blob/master/model_data/coco_classes.txt)
+
+## Postprocessing steps
```
Comment (prasanthpul) [inline yolov3/README.md]: '## -> '###

File: yolov3/README.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+mAP of 0.553 based on original Yolov3 model [here](https://pjreddie.com/darknet/yolo/)
+<hr>
+
+## Publication/Attribution
+Joseph Redmon, Ali Farhadi. YOLOv3: An Incremental Improvement, [paper](https://pjreddie.com/media/files/papers/YOLOv3.pdf)
+
+<hr>
+
+## References
+This model is converted from a keras model [repository](https://github.com/qqwweee/keras-yolo3) using keras2onnx convertor [repository](https://github.com/onnx/keras-onnx).
```
Comment (prasanthpul) [inline yolov3/README.md]: convertor -> converter
